### PR TITLE
Memory backend

### PR DESCRIPTION
**Pull request by Baekalfen**

### CODE REVIEW

**Comment by safl** [comment]
Since ``mem`` is used as the name for one of the backend-interfaces, then a renaming would be nice. Perhaps something along the lines of: ``ramdisk``, ``memdrive``, or something else entirely encapsulating "simple-nvme-device-behavior-emulation-using-virtual-memory-as-storage" :) Or to take inspiration from Linux[1] then ``nulldrive``.
[1] https://www.kernel.org/doc/Documentation/block/null_blk.txt

**Comment by Baekalfen** [comment]
> Since `mem` is used as the name for one of the backend-interfaces, then a renaming would be nice. Perhaps something along the lines of: `ramdisk`, `memdrive`, or something else entirely encapsulating "simple-nvme-device-behavior-emulation-using-virtual-memory-as-storage" :) Or to take inspiration from Linux[1] then `nulldrive`.
> [1] https://www.kernel.org/doc/Documentation/block/null_blk.txt

I chose to rename it to `ramdisk`. Although I like `nulldrive`, it might cause confusion with the `posix_nil` implementation

File: meson_options.txt [hunk 1]
```diff
@@ -2,6 +2,7 @@ option('with-liburing', type: 'boolean', value: true)
 option('with-libaio', type: 'boolean', value: true)
 option('with-spdk', type: 'boolean', value: true)
 option('with-fio', type: 'boolean', value: true)
+option('with-ramdisk', type: 'boolean', value: true)
```
Comment (safl) [inline meson_options.txt:5]: The ``with-x`` has so far been used for things requiring strong third-party dependencies which are retrieved via meson/wrap/subproject. Since ``memdisk`` is an entirely self-contained feature, then it should go into the other options defined below. For example:
```
option('be_ramdisk', type: 'boolean', value: true)
```

File: meson.build [hunk 2]
```diff
@@ -61,6 +61,8 @@ conf_data.set('XNVME_VERSION_PATCH', project_version_patch)
 
 conf_data.set('XNVME_DEBUG_ENABLED', get_option('buildtype') == 'debug')
 
+conf_data.set('XNVME_BE_RAMDISK_ENABLED', get_option('with-ramdisk'))
```
Comment (safl) [inline meson.build:64]: See comment on ``meson_options.txt``

File: include/xnvme_dev.h [hunk 3]
```diff
@@ -12,6 +12,7 @@ enum xnvme_dev_type {
 	XNVME_DEV_TYPE_NVME_NAMESPACE,
 	XNVME_DEV_TYPE_BLOCK_DEVICE,
 	XNVME_DEV_TYPE_FS_FILE,
+	XNVME_DEV_TYPE_RAM_DISK,
```
Comment (safl) [inline include/xnvme_dev.h:15]: See comment about this name further down. TLDR: how about naming it ``XNVME_DEV_TYPE_RAMDISK``?

File: lib/xnvme_be.c [hunk 4]
```diff
@@ -339,6 +344,7 @@ xnvme_be_dev_derive_geometry(struct xnvme_dev *dev)
 	case XNVME_DEV_TYPE_FS_FILE:
 		return _fs_geometry(dev);
 
+	case XNVME_DEV_TYPE_RAM_DISK:
```
Comment (safl) [inline lib/xnvme_be.c:347]: See other comments about ``XNVME_DEV_TYPE_RAM_DISK`` --> ``XNVME_DEV_TYPE_RAMDISK``

File: lib/xnvme_be_ramdisk.c [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+		.mtype = XNVME_BE_ASYNC,
+		.name = "thrpool",
+		.descr = "Use thread pool for Asynchronous I/O",
+		.async = &g_xnvme_be_posix_async_thrpool,
+		.check_support = xnvme_be_supported,
+	},
+
+	{
+		.mtype = XNVME_BE_SYNC,
+		.name = "mem",
```
Comment (safl) [inline lib/xnvme_be_ramdisk.c:31]: Change ``mem`` to ``ramdisk``

File: lib/xnvme_be_ramdisk.c [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+		.mtype = XNVME_BE_ADMIN,
+		.name = "mem",
+		.descr = "Use memory as to construct NVMe idfy responses",
+		.admin = &g_xnvme_be_ramdisk_admin,
+		.check_support = xnvme_be_supported,
+	},
+
+	{
+		.mtype = XNVME_BE_DEV,
+		.name = "mem",
```
Comment (safl) [inline lib/xnvme_be_ramdisk.c:47]: Change ``mem`` to ``ramdisk``

File: lib/xnvme_be_ramdisk.c [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+struct xnvme_be xnvme_be_ramdisk = {
+	.mem = XNVME_BE_NOSYS_MEM,
+	.admin = XNVME_BE_NOSYS_ADMIN,
+	.sync = XNVME_BE_NOSYS_SYNC,
+	.async = XNVME_BE_NOSYS_QUEUE,
+	.dev = XNVME_BE_NOSYS_DEV,
+	.attr =
+		{
+			.name = "mem",
```
Comment (safl) [inline lib/xnvme_be_ramdisk.c:63]: Change ``mem`` to ``ramdisk``

File: Makefile [hunk 8]
```diff
@@ -260,27 +274,27 @@ define gen-bash-completions-help
 # Helper-target to produce Bash-completions for tools (tools, examples, tests)
 #
 # NOTE: This target requires a bunch of things: binaries must be built and
-# residing in 'builddir/tools' etc. AND installed on the system. Also, the find
+# residing in '$(BUILD_DIR)/tools' etc. AND installed on the system. Also, the find
```
Comment (safl) [inline Makefile:277]: I really do prefer that unrelated changes, such as these fixing the hardcoded builddir, would go into a separate commit.

File: Makefile [hunk 9]
```diff
@@ -260,27 +274,27 @@ define gen-bash-completions-help
...
 # command has only been tried with GNU find
 endef
 .PHONY: gen-bash-completions
 gen-bash-completions:
 	@echo "## xNVMe: make gen-bash-completions"
-	$(eval TOOLS := $(shell find builddir/tools builddir/examples builddir/tests -not -name "xnvme_single*" -not -name "xnvme_enum" -not -name "xnvme_dev" -not -name "*.so" -type f -executable -exec basename {} \;))
+	$(eval TOOLS := $(shell find $(BUILD_DIR)/tools $(BUILD_DIR)/examples $(BUILD_DIR)/tests -not -name "xnvme_single*" -not -name "xnvme_enum" -not -name "xnvme_dev" -not -name "*.so" -type f -executable -exec basename {} \;))
```
Comment (safl) [inline Makefile:283]: I really do prefer that unrelated changes, such as these fixing the hardcoded builddir, would go into a separate commit.

File: Makefile [hunk 10]
```diff
@@ -260,27 +274,27 @@ define gen-bash-completions-help
...
 	python3 ./scripts/xnvmec_generator.py cpl --tools ${TOOLS} --output scripts/bash_completion.d
 	@echo "## xNVMe: make gen-bash-completions [DONE]"
 
 define gen-man-pages-help
 # Helper-target to produce man pages for tools (tools, examples, tests)
 #
 # NOTE: This target requires a bunch of things: binaries must be built and
-# residing in 'builddir/tools' etc. AND installed on the system. Also, the find
+# residing in '$(BUILD_DIR)/tools' etc. AND installed on the system. Also, the find
```
Comment (safl) [inline Makefile:291]: I really do prefer that unrelated changes, such as these fixing the hardcoded builddir, would go into a separate commit.

File: Makefile [hunk 11]
```diff
@@ -308,7 +322,7 @@ endef
 .PHONY: tags
 tags:
 	@echo "## xNVMe: make tags"
-	@$(CTAGS) * --languages=C -h=".c.h" -R --exclude=builddir \
+	@$(CTAGS) * --languages=C -h=".c.h" -R --exclude=$(BUILD_DIR) \
```
Comment (safl) [inline Makefile:325]: I really do prefer that unrelated changes, such as these fixing the hardcoded builddir, would go into a separate commit.

File: lib/xnvme_be_ramdisk_sync.c [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+		XNVME_DEBUG("FAILED: nosys opcode: %d", ctx->cmd.common.opcode);
+		return -ENOSYS;
+	}
+
+	return 0;
+}
+#endif
+
+struct xnvme_be_sync g_xnvme_be_ramdisk_sync = {
+	.id = "mem",
```
Comment (safl) [inline lib/xnvme_be_ramdisk_sync.c:43]: Change "mem" to "ramdisk".

File: lib/xnvme_be_ramdisk_dev.c [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+			   xnvme_enumerate_cb XNVME_UNUSED(cb_func), void *XNVME_UNUSED(cb_args))
+{
+	XNVME_DEBUG("FAILED: not implemented (possibly intentional)");
+	return -ENOSYS;
+}
+#endif
+
+struct xnvme_be_dev g_xnvme_be_ramdisk_dev = {
+#ifdef XNVME_BE_RAMDISK_ENABLED
+	.enumerate = xnvme_be_ramdisk_enumerate,
```
Comment (safl) [inline lib/xnvme_be_ramdisk_dev.c:115]: Since there is no enumeration functionality, then set ``.enumerate = xnvme_be_nosys_enumerate`` here and remove enumeration function above.

File: lib/xnvme_be_ramdisk_admin.c [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+
+	default:
+		XNVME_DEBUG("FAILED: ENOSYS opcode: %d", ctx->cmd.common.opcode);
+		return -ENOSYS;
+	}
+}
+#endif
+
+struct xnvme_be_admin g_xnvme_be_ramdisk_admin = {
+	.id = "mem_as_ns",
```
Comment (safl) [inline lib/xnvme_be_ramdisk_admin.c:150]: Change "mem_as_ns" to "ramdisk_as_ns".

**Review by safl — CHANGES_REQUESTED** [review]
See comments inline.

File: Makefile (hunk 11 above)
Comment (Baekalfen) [reply Makefile:311]: Moved to separate commit

File: meson_options.txt (hunk 1 above)
Comment (Baekalfen) [reply meson_options.txt]: Done

File: meson.build (hunk 2 above)
Comment (Baekalfen) [reply meson.build]: Done

File: lib/xnvme_be_ramdisk.c (hunk 5 above)
Comment (Baekalfen) [reply lib/xnvme_be_ramdisk.c]: Done

File: lib/xnvme_be_ramdisk.c [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+		.mtype = XNVME_BE_SYNC,
+		.name = "mem",
+		.descr = "Use memory for synchronous I/O",
+		.sync = &g_xnvme_be_ramdisk_sync,
+		.check_support = xnvme_be_supported,
+	},
+
+	{
+		.mtype = XNVME_BE_ADMIN,
+		.name = "mem",
```
[the remark this answers is not in the document]
Comment (Baekalfen) [reply lib/xnvme_be_ramdisk.c]: Done

File: lib/xnvme_be_ramdisk.c (hunk 7 above)
Comment (Baekalfen) [reply lib/xnvme_be_ramdisk.c]: Done

File: lib/xnvme_be_ramdisk_admin.c [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+	struct xnvme_be_ramdisk_state *state = (void *)dev->be.state;
+	struct xnvme_spec_idfy_ns *ns = dbuf;
+
+	size_t ramdisk_size = _xnvme_be_ramdisk_dev_get_size(dev);
+	if (!ramdisk_size) {
+		return -1;
+	}
+
+	// TODO: Divide by 512 LBA size?
+	ns->nsze = ramdisk_size;
```
[the remark this answers is not in the document]
Comment (Baekalfen) [reply lib/xnvme_be_ramdisk_admin.c]: Fixed to use 512 LBA

File: lib/xnvme_be_ramdisk_sync.c (hunk 12 above)
Comment (Baekalfen) [reply lib/xnvme_be_ramdisk_sync.c]: Done

File: lib/xnvme_be_ramdisk_dev.c (hunk 13 above)
Comment (Baekalfen) [reply lib/xnvme_be_ramdisk_dev.c]: Done, and removed unused function

File: lib/xnvme_be_ramdisk_admin.c (hunk 14 above)
Comment (Baekalfen) [reply lib/xnvme_be_ramdisk_admin.c]: Done

File: lib/xnvme_be_ramdisk_admin.c [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+	struct xnvme_be_ramdisk_state *state = (void *)dev->be.state;
+	struct xnvme_spec_fs_idfy_ns *ns = dbuf;
+
+	size_t ramdisk_size = _xnvme_be_ramdisk_dev_get_size(dev);
+	if (!ramdisk_size) {
+		return -1;
+	}
+
+	// TODO: Divide by 512 LBA size?
+	ns->nsze = ramdisk_size;
```
[the remark this answers is not in the document]
Comment (Baekalfen) [reply lib/xnvme_be_ramdisk_admin.c:46]: Done

**Review by safl — APPROVED** [review]

File: meson.build [hunk 18]
```diff
@@ -61,6 +61,8 @@ conf_data.set('XNVME_VERSION_PATCH', project_version_patch)
 
 conf_data.set('XNVME_DEBUG_ENABLED', get_option('buildtype') == 'debug')
 
+conf_data.set('XNVME_BE_RAMDISK_ENABLED', get_option('be_ramdisk'))
```
Comment (safl) [inline meson.build:64]: Great!

File: lib/xnvme_be_ramdisk_sync.c [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+
+int
+xnvme_be_ramdisk_sync_cmd_io(struct xnvme_cmd_ctx *ctx, void *dbuf, size_t dbuf_nbytes,
+			     void *XNVME_UNUSED(mbuf), size_t XNVME_UNUSED(mbuf_nbytes))
+{
+	struct xnvme_be_ramdisk_state *state = (void *)ctx->dev->be.state;
+	const uint64_t ssw = ctx->dev->geo.ssw;
+
+	switch (ctx->cmd.common.opcode) {
+	case XNVME_SPEC_NVM_OPC_WRITE:
```
Comment (safl) [inline lib/xnvme_be_ramdisk_sync.c:23]: You can add handling of the pseudo-fs-spec here, it is the same as for NVM but without the need to shift by ``ssw``.

File: lib/xnvme_be_ramdisk_dev.c [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+	memset(&dev->be, 0, sizeof(dev->be));
+}
+
+size_t
+_xnvme_be_ramdisk_dev_get_size(struct xnvme_dev *dev)
+{
+	struct xnvme_be_ramdisk_state *state = (void *)dev->be.state;
+	const struct xnvme_ident *ident = &dev->ident;
+
+	char *gb_marker = strrchr(ident->uri, 'G');
```
Comment (safl) [inline lib/xnvme_be_ramdisk_dev.c:44]: See previous comments on moving up declaration and not doing logic which requires error-handling.

**Review by safl — CHANGES_REQUESTED** [review]
See requests-for-changes inline.

Additionally, regarding the ``sync`` implementation; a definition of ``.cmd_iov()`` is missing, either assign ``xnvme_be_nosys_sync_cmd_iov()``, or **preferably** add an actual implementation handling ``iovec`` payloads.

File: lib/xnvme_be.c [hunk 21]
```diff
@@ -698,6 +705,7 @@ xnvme_be_factory(struct xnvme_dev *dev, struct xnvme_opts *opts)
 			continue;
 		}
 
+		XNVME_DEBUG("INFO: checking be: '%s'", be.attr.name);
```
[the remark this answers is not in the document]
Comment (Baekalfen) [reply lib/xnvme_be.c:708]: This one was superfluous.

File: lib/xnvme_be.c [hunk 22]
```diff
@@ -688,6 +694,7 @@ xnvme_be_factory(struct xnvme_dev *dev, struct xnvme_opts *opts)
 	for (int i = 0; (i < g_xnvme_be_count) && g_xnvme_be_registry[i]; ++i) {
 		struct xnvme_be be = *g_xnvme_be_registry[i];
 		int setup = 0;
+		XNVME_DEBUG("INFO: checking be: '%s'", be.attr.name);
```
[the remark this answers is not in the document]
Comment (Baekalfen) [reply lib/xnvme_be.c:697]: We only have log-statements on failures, which makes it hard to find the context in the logs.

**Review by safl — APPROVED** [review]

File: lib/xnvme_be_ramdisk_sync.c [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+	return 0;
+}
+#endif
+
+struct xnvme_be_sync g_xnvme_be_ramdisk_sync = {
+	.id = "ramdisk",
+#ifdef XNVME_BE_RAMDISK_ENABLED
+	.cmd_io = xnvme_be_ramdisk_sync_cmd_io,
+#else
+	.cmd_io = xnvme_be_nosys_sync_cmd_io,
```
Comment (safl) [inline lib/xnvme_be_ramdisk_sync.c:59]: Assign the enosys implementation of ``cmd_iov()`` here.

Just a thought: this might be inspiration for slighty more tricky thing that a ``pre-commit`` hook could help catch, unassigned backend interface implementations.

File: lib/xnvme_be_ramdisk_dev.c [hunk 24]
```diff
@@ -0,0 +1,116 @@
+// re
```
Comment (safl) [inline lib/xnvme_be_ramdisk_dev.c:1]: Typo?

**Review by safl — CHANGES_REQUESTED** [review]
Thank you @Baekalfen, this is looking really good!
Added a couple of comments in-line.

Additionally, then it needs some verification and documentation. On testing, this would be adding a testplan exercising the backend with a fitting set of testcases. This would be done by expanding the ``cijoe-pkg-xnvme``. However, this is a bit cumbersome as the changes must be added to an entirely different repository. To address that, this is the way to go: https://github.com/OpenMPDK/xNVMe/issues/44 with needs a bit of attention, this is needed for a bunch of other things. So, I'll take a look at that tomorrow.

On the documentation part, then all backends need more of that... so for now just add a quick note, in the commit-message, on the ``dev_uri`` which the backend uses.

File: lib/xnvme_be_ramdisk_sync.c (hunk 23 above)
Comment (Baekalfen) [reply lib/xnvme_be_ramdisk_sync.c:59]: Added. Otherwise, let nosys be implicit.

A `pre-commit` hook might be able to infer it from a symbols table or some other preexisting build result.

File: lib/xnvme_be_ramdisk_dev.c (hunk 24 above)
Comment (Baekalfen) [reply lib/xnvme_be_ramdisk_dev.c:1]: Removed

File: lib/xnvme_be_ramdisk_sync.c [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	return 0;
+}
+#endif
+
+struct xnvme_be_sync g_xnvme_be_ramdisk_sync = {
+	.id = "ramdisk",
+#ifdef XNVME_BE_RAMDISK_ENABLED
+	.cmd_io = xnvme_be_ramdisk_sync_cmd_io,
```
[the remark this answers is not in the document]
Comment (Baekalfen) [reply lib/xnvme_be_ramdisk_sync.c:57]: I added `nosys` for now. Let's get the PR through, and then incrementally improve afterwards.

**Comment by safl** [comment]
Find the rebased branch here: https://github.com/OpenMPDK/xNVMe/tree/for/ghpr-38
The commit ``build:make: change hardcoded builddir to variable BUILD_DIR`` is now on next.

**Comment by Baekalfen** [comment]
@safl If you have no more comments, then I consider this PR ready for merge.

File: lib/xnvme_be_ramdisk_dev.c [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+xnvme_be_ramdisk_dev_open(struct xnvme_dev *dev)
+{
+	struct xnvme_be_ramdisk_state *state = (void *)dev->be.state;
+	const struct xnvme_ident *ident = &dev->ident;
+	struct xnvme_opts *opts = &dev->opts;
+	int err;
+
+	size_t ramdisk_size = _xnvme_be_ramdisk_dev_get_size(dev);
+	if (!ramdisk_size) {
+		return -2;
```
Comment (safl) [inline lib/xnvme_be_ramdisk_dev.c:63]: This should be an `errno` const.

File: lib/xnvme_be_ramdisk_sync.c [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+int
+xnvme_be_ramdisk_sync_cmd_io(struct xnvme_cmd_ctx *ctx, void *dbuf, size_t dbuf_nbytes,
+			     void *XNVME_UNUSED(mbuf), size_t XNVME_UNUSED(mbuf_nbytes))
+{
+	struct xnvme_be_ramdisk_state *state = (void *)ctx->dev->be.state;
+	const uint64_t ssw = ctx->dev->geo.ssw;
+
+	switch (ctx->cmd.common.opcode) {
+	case XNVME_SPEC_NVM_OPC_WRITE:
+		memcpy((char *)(state->ramdisk) + (size_t)(ctx->cmd.nvm.slba << ssw), dbuf,
```
Comment (safl) [inline lib/xnvme_be_ramdisk_sync.c:24]: How about declaring a:
```
char * offset = state->ramdisk;
```
and using that in the memcpy call and drop the cast to ``size-t``:
```
memcpy(offset + (ctx->cmd.nvm.slba << ssw), ...)
```
?

**Review by safl — CHANGES_REQUESTED** [review]
Added comments inline,  other than that then it would be good with a commit adding the testplan to the Ci. Since there are no NVMe / device-requirements, then this could be added as an extension to the build-test, providing an actual exercise of xNVMe on the various systems.
